system/src/control/storage.cpp

```diff
@@ -0,0 +1,496 @@
+#endif
+
+// TODO: Add support for external flash (available if HAS_SERIAL_FLASH macro is defined)
+#ifdef USE_SERIAL_FLASH
+#error "External flash is not supported"
+#endif
+
+namespace particle {
+
+namespace control {
+
+using namespace protocol;
+using namespace common;
+
+namespace {
+
+struct Section;
+
+// Description of a storage
+struct Storage {
+    uint8_t type;
+    uint8_t flags;
+    const Section* sections;
+    uint8_t sectionCount;
+};
+
+// IO callbacks
+typedef int(*sectionRead)(char* data, size_t size, uintptr_t address);
+typedef int(*sectionWrite)(const char* data, size_t size, uintptr_t address);
+typedef int(*sectionClear)(uintptr_t address, size_t size);
+typedef int(*sectionGetSize)(uintptr_t address, size_t size, size_t& dataSize);
+
+// Description of a storage section
+struct Section {
+    uint8_t type;
+    uint8_t firmwareModuleType;
+    uint8_t firmwareModuleIndex;
+    uint8_t flags;
+    uint32_t address;
+    uint32_t size;
+    sectionRead read;
+    sectionWrite write;
+    sectionClear clear;
+    sectionGetSize getSize;
+};
+
+// Internal flash
+int internalFlashRead(char* data, size_t size, uintptr_t address) {
+    return InternalFlashStore::read(address, data, size);
+}
+
+int internalFlashWrite(const char* data, size_t size, uintptr_t address) {
+    return InternalFlashStore::write(address, data, size);
+}
+
+int internalFlashClear(uintptr_t address, size_t size) {
+    return InternalFlashStore::erase(address, size);
+}
+
+int getFirmwareModuleSize(uintptr_t address, size_t maxSize, size_t& moduleSize) {
+    const module_info_t* module = FLASH_ModuleInfo(FLASH_INTERNAL, address);
+    if (!module) {
+        return SYSTEM_ERROR_UNKNOWN;
+    }
+    // Check module boundaries
+    const uintptr_t startAddr = (uintptr_t)module->module_start_address;
+    const uintptr_t endAddr = (uintptr_t)module->module_end_address;
+    const size_t size = endAddr - startAddr;
+    if (endAddr < startAddr || size > maxSize) {
+        return SYSTEM_ERROR_BAD_DATA;
+    }
+    // Verify checksum
+    if (!FLASH_VerifyCRC32(FLASH_INTERNAL, address, size)) {
+        return SYSTEM_ERROR_BAD_DATA;
+    }
+    moduleSize = size + 4; // Include CRC
+    return 0;
+}
+
+// DCT/DCD
+int configRead(char* data, size_t size, uintptr_t address) {
+    return dct_read_app_data_copy(address, data, size);
+}
+
+int configWrite(const char* data, size_t size, uintptr_t address) {
+    return dct_write_app_data(data, address, size);
+}
+
+// EEPROM
+int eepromRead(char* data, size_t size, uintptr_t address) {
+    HAL_EEPROM_Get(address, data, size); // FIXME: Result code?
+    return 0;
+}
+
+int eepromWrite(const char* data, size_t size, uintptr_t address) {
+    HAL_EEPROM_Put(address, data, size); // FIXME: Result code?
+    return 0;
+}
+
+int eepromClear(uintptr_t address, size_t size) {
+    HAL_EEPROM_Clear(); // FIXME: Result code?
+    return 0;
+}
+
+const Section INTERNAL_STORAGE_SECTIONS[] = {
```

Comment:
Could we add a storage for the entire flash space? It's low level, but might be useful in some situations.